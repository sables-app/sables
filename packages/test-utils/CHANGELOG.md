# @sables-app/test-utils

## 0.1.2

### Patch Changes

- 5460928: Publish dev packages to public registry install of GitHub Packages
- Updated dependencies [5460928]
  - @sables-app/scripts@0.1.4

## 0.1.1

### Patch Changes

- 776123f: chore(deps): Bump vitest from 0.26.2 to 0.26.3
