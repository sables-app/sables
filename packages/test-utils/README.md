# `@sables-app/test-utils`

> Sables test utilities.

## Install

```sh
npm install --save-dev @sables-app/test-utils
```
