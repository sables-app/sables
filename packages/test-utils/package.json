{
  "name": "@sables-app/test-utils",
  "version": "0.1.1",
  "description": "Sables test utilities.",
  "author": "Morris Allison III (https://morris.xyz)",
  "license": "MIT",
  "bugs": "https://github.com/sables-app/sables/issues",
  "homepage": "https://github.com/sables-app/sables/tree/main/packages/test-utils",
  "repository": {
    "type": "git",
    "url": "https://github.com/sables-app/sables.git",
    "directory": "packages/test-utils"
  },
  "type": "module",
  "main": "./dist/main.js",
  "dependencies": {
    "@sables-app/scripts": "~0.1.0",
    "@types/node": "^18.11.18",
    "vitest": "~0.26.3"
  },
  "devDependencies": {
    "typescript": "^4.9.3"
  },
  "files": [
    "dist",
    "!__tests__"
  ]
}
