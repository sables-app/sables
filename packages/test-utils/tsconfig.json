{
  "extends": "@sables-app/scripts/tsconfig.package.json",
  "include": ["src/**/*", "src/**/*.json"],
  "exclude": ["__mocks__", "__snapshots__", "dist", "node_modules"],
  "compilerOptions": {
    "module": "es2022",
    "noEmit": false,
    "outDir": "dist",
    "rootDir": "src"
  }
}
