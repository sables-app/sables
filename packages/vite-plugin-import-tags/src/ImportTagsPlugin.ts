import { createFilter, FilterPattern } from "@rollup/pluginutils";
import crypto from "crypto";
import fs from "fs-jetpack";
import MagicString from "magic-string";
import path from "path";
import type { RenderChunkHook, TransformHook } from "rollup";
import type { ConfigEnv, UserConfig } from "vite";

import type { ImportTagInfo, ImportTagsManifest } from "./types.js";
import { IMPORT_TAGS_FILENAME } from "./utils.js";

interface PluginConfig {
  exclude?: FilterPattern;
  include?: FilterPattern;
  manifest: boolean;
  sourceMap: boolean;
}

const DEFAULT_OPTIONS: PluginConfig = {
  manifest: true,
  sourceMap: true,
};

const IMPORT_PATTERN =
  /(import\(("|')([a-zA-Z0-9-_@./]+)("|')\))((\.then|\.catch|\.finally))*(?!(\[|\.))/g;

function logInfo(message: string) {
  console.info(`[ImportTagsPlugin]: ${message}`);
}

export function createImportTagsPlugin(options?: Partial<PluginConfig>) {
  const config: PluginConfig = { ...DEFAULT_OPTIONS, ...options };
  const filter = createFilter(config.include, config.exclude);
  const manifest: ImportTagsManifest = {};

  const cwd = process.cwd();
  let buildDir = cwd;
  let pluginEnabled = true;
  let ssrEntryDir = path.dirname(cwd);

  function idToSrc(id: string) {
    if (path.isAbsolute(id) && fs.exists(id)) {
      return path.relative(ssrEntryDir, id);
    }
    return id;
  }

  function codeHasReplacements(
    code: string,
    id: string,
    magicString: MagicString
  ) {
    let result = false;
    let matches: RegExpExecArray | null;

    while ((matches = IMPORT_PATTERN.exec(code))) {
      result = true;

      const start = matches.index;
      const [, importMatch, , importPath] = matches;
      const end = start + importMatch.length;
      const src = idToSrc(id);
      const entry: ImportTagInfo = { src, importPath };
      const hash = crypto
        .createHash("md5")
        .update(JSON.stringify(entry))
        .digest("hex");
      const replacement = `($rollupImportTag("${hash}"),import("${importPath}"))`;
      magicString.overwrite(start, end, replacement);
      manifest[hash] = entry;
    }

    return result;
  }

  function executeReplacement(
    code: string,
    id: string
  ): ReturnType<RenderChunkHook> {
    const magicString = new MagicString(code);

    if (!codeHasReplacements(code, id, magicString)) {
      return null;
    }

    return {
      code: magicString.toString(),
      map: config.sourceMap
        ? magicString.generateMap({ hires: true })
        : undefined,
    };
  }

  function transform(
    ...params: Parameters<TransformHook>
  ): ReturnType<TransformHook> {
    if (!pluginEnabled) return;

    const [code, id] = params;

    if (!filter(id)) return null;

    return executeReplacement(code, id);
  }

  function configHook(config: UserConfig, env: ConfigEnv): void {
    if (config.build?.outDir) {
      buildDir = config.build.outDir;
    }
    if (typeof config.build?.ssr == "string") {
      ssrEntryDir = path.dirname(config.build.ssr);
    }

    pluginEnabled = env.command === "build" && !!env.ssrBuild;
  }

  function buildEnd(error?: Error): void {
    if (!pluginEnabled) return;

    logInfo(`${Object.keys(manifest).length} imports were tagged.`);

    if (!error && buildDir && config.manifest) {
      logInfo("Writing manifest to output directory.");

      const manifestFilePath = path.resolve(buildDir, IMPORT_TAGS_FILENAME);

      fs.write(manifestFilePath, manifest);
    }
  }

  return {
    buildEnd,
    config: configHook,
    name: "importTags",
    transform,
  };
}
