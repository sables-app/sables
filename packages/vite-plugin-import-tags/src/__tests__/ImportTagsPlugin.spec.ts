import "./setupMocks.js";

import { silenceLogs } from "@sables-app/test-utils";

import path from "node:path";

import fs from "fs-jetpack";
import { beforeEach, describe, expect, it } from "vitest";
import * as vitest from "vitest";

import { createImportTagsPlugin } from "../ImportTagsPlugin.js";

const codeFixture = `
import "moduleA";
import { value } from "moduleB";

const foo = "bar";

function hello() {
  return import("./moduleC");
}

/** @type {import('my-package')['someProperty']} */
import("./moduleD").then(console.log);
import("./moduleF").catch(console.log);
import("./moduleG").finally(console.log);

/**
 * import('my-package').someProperty
 * import('my-package')['someProperty']
 * \`import('my-package').someProperty\`
 * \`import('my-package')['someProperty']\`
 */
const qux = (await import("./moduleE")).default;

/**
 * Sets focus at a given node. The last focused element will help to determine which element(first or last) should be focused.
 * HTML markers (see {@link import('./constants').FOCUS_AUTO} constants) can control autofocus
 * @param topNode
 * @param lastNode
 * @param options
 */
`;

describe("ImportTagsPlugin", () => {
  silenceLogs(vitest, "info");

  describe("plugin", () => {
    const fsWriteMock = fs.write as vitest.Mock;

    let plugin: ReturnType<typeof createImportTagsPlugin>;

    beforeEach(() => {
      fsWriteMock.mockReset();
      plugin = createImportTagsPlugin();
    });

    function invokeTransform() {
      return plugin.transform(codeFixture, "./myModules/MyModule.tsx");
    }

    describe("transform", () => {
      it("transforms modules", () => {
        const result = invokeTransform();
        const code = typeof result == "string" ? result : result?.code;

        expect(code).toMatchSnapshot();
      });
    });

    describe("buildEnd", () => {
      beforeEach(() => {
        invokeTransform();
      });

      it("writes a manifest to the build directory", () => {
        expect(fsWriteMock).toHaveBeenCalledTimes(0);

        plugin.buildEnd();

        expect(fsWriteMock).toHaveBeenCalledOnce();

        const [manifestPath, manifest] = fsWriteMock.mock.calls[0];
        const expectedManifestPathSuffix = path.resolve(
          process.cwd(),
          "importTagsManifest.json"
        );

        expect(path.isAbsolute(manifestPath)).toBe(true);
        expect(manifestPath).toBe(expectedManifestPathSuffix);
        expect(manifest).toMatchSnapshot();
      });

      describe("when an error occurs", () => {
        it("doesn't write the manifest", () => {
          expect(fsWriteMock).toHaveBeenCalledTimes(0);

          plugin.buildEnd(new Error());

          expect(fsWriteMock).toHaveBeenCalledTimes(0);
        });
      });
    });
  });
});
