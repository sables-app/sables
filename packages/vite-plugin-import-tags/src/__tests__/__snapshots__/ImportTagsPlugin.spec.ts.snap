// Vitest Snapshot v1

exports[`ImportTagsPlugin > plugin > buildEnd > writes a manifest to the build directory 1`] = `
{
  "6515d76aa069b88f3b2961aeab57f1e9": {
    "importPath": "./moduleG",
    "src": "./myModules/MyModule.tsx",
  },
  "677cd3a887eb7012ef48e37ac2cf09cf": {
    "importPath": "./moduleD",
    "src": "./myModules/MyModule.tsx",
  },
  "857881fb373d96ddb418e2f25557185c": {
    "importPath": "./moduleF",
    "src": "./myModules/MyModule.tsx",
  },
  "c6964f1e4bad9154b92913eecd5094dd": {
    "importPath": "./moduleE",
    "src": "./myModules/MyModule.tsx",
  },
  "d1aa1b2ebc20e39160a6e54e1e1cc7dc": {
    "importPath": "./moduleC",
    "src": "./myModules/MyModule.tsx",
  },
}
`;

exports[`ImportTagsPlugin > plugin > transform > transforms modules 1`] = `
"
import \\"moduleA\\";
import { value } from \\"moduleB\\";

const foo = \\"bar\\";

function hello() {
  return ($rollupImportTag(\\"d1aa1b2ebc20e39160a6e54e1e1cc7dc\\"),import(\\"./moduleC\\"));
}

/** @type {import('my-package')['someProperty']} */
($rollupImportTag(\\"677cd3a887eb7012ef48e37ac2cf09cf\\"),import(\\"./moduleD\\")).then(console.log);
($rollupImportTag(\\"857881fb373d96ddb418e2f25557185c\\"),import(\\"./moduleF\\")).catch(console.log);
($rollupImportTag(\\"6515d76aa069b88f3b2961aeab57f1e9\\"),import(\\"./moduleG\\")).finally(console.log);

/**
 * import('my-package').someProperty
 * import('my-package')['someProperty']
 * \`import('my-package').someProperty\`
 * \`import('my-package')['someProperty']\`
 */
const qux = (await ($rollupImportTag(\\"c6964f1e4bad9154b92913eecd5094dd\\"),import(\\"./moduleE\\"))).default;

/**
 * Sets focus at a given node. The last focused element will help to determine which element(first or last) should be focused.
 * HTML markers (see {@link import('./constants').FOCUS_AUTO} constants) can control autofocus
 * @param topNode
 * @param lastNode
 * @param options
 */
"
`;
